### PR TITLE
Adding openstack bakery (#3)

### PR DESCRIPTION
**Pull request by edwinavalos**
Adding Openstack Bakery commands to halyard

h/t @shazy792 @dylanbernhardt @matthewkohner @msilpala

### CODE REVIEW

File: halyard-cli/src/main/java/com/netflix/spinnaker/halyard/cli/command/v1/config/providers/openstack/OpenstackAddBaseImageCommand.java [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+
+import com.beust.jcommander.Parameter;
+import com.beust.jcommander.Parameters;
+import com.netflix.spinnaker.halyard.cli.command.v1.config.providers.bakery.AbstractAddBaseImageCommand;
+import com.netflix.spinnaker.halyard.config.model.v1.node.BaseImage;
+import com.netflix.spinnaker.halyard.config.model.v1.providers.openstack.OpenstackBaseImage;
+
+@Parameters(separators = "=")
+public class OpenstackAddBaseImageCommand extends AbstractAddBaseImageCommand{
+    protected String getProviderName() {return "openstack"; }
```
Comment (lwander) [inline halyard-cli/src/main/java/com/netflix/spinnaker/halyard/cli/command/v1/config/providers/openstack/OpenstackAddBaseImageCommand.java]: Formatting seems off here

File: halyard-cli/src/main/java/com/netflix/spinnaker/halyard/cli/command/v1/config/providers/openstack/OpenstackEditAccountCommand.java [hunk 2]
```diff
@@ -75,8 +87,32 @@ protected String getProviderName() {
...
+      names = "--consul-config",
+      description = OpenstackCommandProperties.CONSUL_CONFIG_DESCRIPTION
+  )
+  private String consulConfig;
+
+  @Parameter(
+      names = "--remove-consul-config",
+      description = "Removes currently configured consul config file."
+  )
+  private boolean removeConsulConfig;
```
Comment (lwander) [inline halyard-cli/src/main/java/com/netflix/spinnaker/halyard/cli/command/v1/config/providers/openstack/OpenstackEditAccountCommand.java]: These "-remove" style flags aren't needed. You can always clear with `--consul-config=`.

File: halyard-config/src/main/java/com/netflix/spinnaker/halyard/config/model/v1/providers/openstack/OpenstackProvider.java [hunk 3]
```diff
@@ -16,4 +20,19 @@ public ProviderType providerType() {
...
   }
+
+  @Override
+  public  OpenstackBakeryDefaults emptyBakeryDefaults() {
+    OpenstackBakeryDefaults result = new OpenstackBakeryDefaults();
+    // Set defaults over here
+    result.setAuthUrl("http://default.com");
+    result.setFloatingIpPool("default_pool");
+    result.setInsecure(false);
+    result.setNetworkId("xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx");
```
Comment (lwander) [inline halyard-config/src/main/java/com/netflix/spinnaker/halyard/config/model/v1/providers/openstack/OpenstackProvider.java]: These defaults will be set any time someone initializes the bakery options. The way these read, I'm guessing they'd have to be cleared/overwritten each time for this to be useful. If that's the case, just leave the fields blank.

File: halyard-config/src/main/java/com/netflix/spinnaker/halyard/config/validate/v1/providers/openstack/OpenstackAccountValidator.java [hunk 4]
```diff
@@ -1,54 +1,91 @@
...
     String projectName = account.getPassword();
     String domainName = account.getDomainName();
+    String authUrl = account.getAuthUrl();
+    List<String> regions = account.getRegions();
     Boolean insecure = account.getInsecure();
-    String userDataFile = account.getUserDataFile();
+    String heatTemplateLocation = account.getHeatTemplateLocation();
     OpenstackAccount.OpenstackLbaasOptions lbaas = account.getLbaas();
-    String regions = account.getRegions();
+    // TODO: consul config does not yet exist for OpenStack
```
Comment (lwander) [inline halyard-config/src/main/java/com/netflix/spinnaker/halyard/config/validate/v1/providers/openstack/OpenstackAccountValidator.java]: Where does it not exist? Could you link to an open bug in the component where it needs to be added in the `TODO()` braces?

File: halyard-config/src/main/java/com/netflix/spinnaker/halyard/config/validate/v1/providers/openstack/OpenstackAccountValidator.java [hunk 5]
```diff
@@ -1,54 +1,91 @@
...
 
+    if (projectName == null || projectName.isEmpty()) {
+      psBuilder.addProblem(Problem.Severity.ERROR, "You must provide a project name");
+    }
+
     if (!StringUtils.endsWith(authUrl, "/v3")) {
       psBuilder.addProblem(Problem.Severity.WARNING, "You must use Keystone v3. The default auth url will be of the format IP:5000/v3.");
     }
 
+    if (domainName == null || domainName.isEmpty()) {
```
Comment (lwander) [inline halyard-config/src/main/java/com/netflix/spinnaker/halyard/config/validate/v1/providers/openstack/OpenstackAccountValidator.java]: All of these string checks can be replaced with `StringUtils.isEmpty(..)` from `org.apache.commons.lang3`.

File: halyard-config/src/main/java/com/netflix/spinnaker/halyard/config/validate/v1/providers/openstack/OpenstackBakeryDefaultsValidator.java [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+                StringUtils.isEmpty(securityGroups) &&
+                StringUtils.isEmpty(projectName) &&
+                StringUtils.isEmpty(username) &&
+                StringUtils.isEmpty(password) &&
+                StringUtils.isEmpty(networkId) &&
+                CollectionUtils.isEmpty(baseImages)) {
+            return;
+        }
+
+
```
Comment (lwander) [inline halyard-config/src/main/java/com/netflix/spinnaker/halyard/config/validate/v1/providers/openstack/OpenstackBakeryDefaultsValidator.java]: Extra newline

File: halyard-config/src/main/java/com/netflix/spinnaker/halyard/config/validate/v1/providers/openstack/OpenstackBakeryDefaultsValidator.java [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+        String networkId = n.getNetworkId();
+        String floatingIpPool = n.getFloatingIpPool();
+        String securityGroups = n.getSecurityGroups();
+        String projectName = n.getProjectName();
+        String username = n.getUsername();
+        String password = n.getPassword();
+        boolean insecure = n.isInsecure();
+        List<OpenstackBaseImage> baseImages = n.getBaseImages();
+
+        if (StringUtils.isEmpty(authUrl) &&
```
Comment (lwander) [inline halyard-config/src/main/java/com/netflix/spinnaker/halyard/config/validate/v1/providers/openstack/OpenstackBakeryDefaultsValidator.java]: Why is this acceptable?

File: halyard-config/src/main/java/com/netflix/spinnaker/halyard/config/validate/v1/providers/openstack/OpenstackBaseImageValidator.java [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+
+        if (StringUtils.isEmpty(sourceImageId)) {
+            p.addProblem(Problem.Severity.ERROR, "No source image id supplied for openstack base image.");
+        }
+
+        if (StringUtils.isEmpty(sshUserName)) {
+            p.addProblem(Problem.Severity.ERROR, "No ssh username supplied for openstack base image.");
+        }
+
+        // TODO(shazy792) Add check to see if image actually exists on openstack instance(thum
```
Comment (lwander) [inline halyard-config/src/main/java/com/netflix/spinnaker/halyard/config/validate/v1/providers/openstack/OpenstackBaseImageValidator.java]: Looks like the comment got cut off here

File: halyard-config/src/main/java/com/netflix/spinnaker/halyard/config/validate/v1/providers/openstack/OpenstackProviderValidator.java [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+    @Override
+    public void validate(ConfigProblemSetBuilder p, OpenstackProvider n) {
+        List<OpenstackNamedAccountCredentials> credentialsList = new ArrayList<>();
+
+        OpenstackAccountValidator openstackAccountValidator = new OpenstackAccountValidator(credentialsList, halyardVersion);
+
+        n.getAccounts().forEach(openstackAccount -> openstackAccountValidator.validate(p, openstackAccount));
+
+        new OpenstackBakeryDefaultsValidator(credentialsList, halyardVersion).validate(p, n.getBakeryDefaults());
+
```
Comment (lwander) [inline halyard-config/src/main/java/com/netflix/spinnaker/halyard/config/validate/v1/providers/openstack/OpenstackProviderValidator.java]: Extra newline

File: halyard-config/src/main/java/com/netflix/spinnaker/halyard/config/validate/v1/providers/openstack/OpenstackBakeryDefaultsValidator.java (hunk 7 above)
Comment (dylanbernhardt) [inline halyard-config/src/main/java/com/netflix/spinnaker/halyard/config/validate/v1/providers/openstack/OpenstackBakeryDefaultsValidator.java]: We kept this line so the bakery defaults can be null when an account is enabled and added, but bakery parameters aren't set yet.

Comment (lwander) [inline halyard-config/src/main/java/com/netflix/spinnaker/halyard/config/validate/v1/providers/openstack/OpenstackBakeryDefaultsValidator.java]: Ah, and there are no defaults that can be set which work out of the box?

Comment (dylanbernhardt) [inline halyard-config/src/main/java/com/netflix/spinnaker/halyard/config/validate/v1/providers/openstack/OpenstackBakeryDefaultsValidator.java]: correct

Comment (lwander) [inline halyard-config/src/main/java/com/netflix/spinnaker/halyard/config/validate/v1/providers/openstack/OpenstackBakeryDefaultsValidator.java]: Ok makes sense, thanks

File: halyard-cli/src/main/java/com/netflix/spinnaker/halyard/cli/command/v1/config/providers/openstack/OpenstackEditAccountCommand.java [hunk 10]
```diff
@@ -75,20 +87,26 @@ protected String getProviderName() {
...
+      names = "--heat-template-location",
       converter = PathExpandingConverter.class,
-      description = OpenstackCommandProperties.USER_DATA_FILE_DESCRIPTION
+      description = OpenstackCommandProperties.HEAT_TEMPLATE_LOCATION_DESCRIPTION
   )
-  private String userDataFile;
+  private String heatTemplateLocation;
 
   @Parameter(
-      names = "--remove-user-data-file",
```
Comment (markus-silpala-tgt) [inline halyard-cli/src/main/java/com/netflix/spinnaker/halyard/cli/command/v1/config/providers/openstack/OpenstackEditAccountCommand.java]: Did we deliberately remove this parameter? @edwinavalos

File: halyard-cli/src/main/java/com/netflix/spinnaker/halyard/cli/command/v1/config/providers/openstack/OpenstackBaseImageCommand.java [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.netflix.spinnaker.halyard.cli.command.v1.config.providers.openstack;
+
+
+import com.beust.jcommander.Parameters;
+import com.netflix.spinnaker.halyard.cli.command.v1.config.providers.bakery.AbstractBaseImageCommand;
+
```
Comment (markus-silpala-tgt) [inline halyard-cli/src/main/java/com/netflix/spinnaker/halyard/cli/command/v1/config/providers/openstack/OpenstackBaseImageCommand.java]: Most of our commands seem to have a little class-level comment describing what they do in very generic terms. Do we want that here? @edwinavalos

File: halyard-cli/src/main/java/com/netflix/spinnaker/halyard/cli/command/v1/config/providers/openstack/OpenstackEditAccountCommand.java (hunk 10 above)
Comment (edwinavalos) [inline halyard-cli/src/main/java/com/netflix/spinnaker/halyard/cli/command/v1/config/providers/openstack/OpenstackEditAccountCommand.java]: Yep, Lars explained a difference between these removes and others in halyard,only removes that edit a list should be implemented the halyard way is to set the variable to nothing

File: halyard-cli/src/main/java/com/netflix/spinnaker/halyard/cli/command/v1/config/providers/openstack/OpenstackCommandProperties.java [hunk 12]
```diff
@@ -17,7 +21,19 @@
...
 
   static final String LBAAS_POLL_INTERVAL_DESCRIPTION = "Interval in seconds to poll octavia when an entity is created, updated, or deleted. Default 5.";
+
+  static final String INSTANCE_TYPE_DESCRIPTION = "The instance type for the baking configuration.";
+
+  static final String SOURCE_IMAGE_ID_DESCRIPTION = "The source image ID for the baking configuration.";
+
+  static final String SSH_USER_NAME_DESCRIPTION = "The ssh username for the baking configuration.";
+
+  static final String REGION_DESCRIPTION = "The region for the baking configuration.";
```
Comment (markus-silpala-tgt) [inline halyard-cli/src/main/java/com/netflix/spinnaker/halyard/cli/command/v1/config/providers/openstack/OpenstackCommandProperties.java]: @edwinavalos Does this argument support multiple regions separated by commas? If so it's likely worth calling that  out here in the description. If not, thank you for educating me.

File: halyard-config/src/main/java/com/netflix/spinnaker/halyard/config/model/v1/providers/openstack/OpenstackAccount.java [hunk 13]
```diff
@@ -1,25 +1,36 @@
...
+  private String accountType;
   private String authUrl;
   private String username;
   private String password;
   private String projectName;
   private String domainName;
   private Boolean insecure = false;
+  private String heatTemplateLocation;
+  private String consulConfig;
+//  private  lbaasConfig;
```
Comment (markus-silpala-tgt) [inline halyard-config/src/main/java/com/netflix/spinnaker/halyard/config/model/v1/providers/openstack/OpenstackAccount.java]: Any reason not to just delete this line?

File: halyard-config/src/main/java/com/netflix/spinnaker/halyard/config/model/v1/providers/openstack/OpenstackProvider.java [hunk 14]
```diff
@@ -1,11 +1,15 @@
 package com.netflix.spinnaker.halyard.config.model.v1.providers.openstack;
 
+import com.netflix.spinnaker.halyard.config.model.v1.node.HasImageProvider;
 import com.netflix.spinnaker.halyard.config.model.v1.node.Provider;
```
Comment (markus-silpala-tgt) [inline halyard-config/src/main/java/com/netflix/spinnaker/halyard/config/model/v1/providers/openstack/OpenstackProvider.java]: Unused import.

File: halyard-config/src/main/java/com/netflix/spinnaker/halyard/config/services/v1/BakeryService.java [hunk 15]
```diff
@@ -77,7 +77,8 @@ private BaseImage getBaseImage(NodeFilter filter, String baseImageName) {
     }
   }
 
-  public BakeryDefaults getBakeryDefaults(String deploymentName, String providerName) {
+  public BakeryDefaults
+  getBakeryDefaults(String deploymentName, String providerName) {
```
Comment (markus-silpala-tgt) [inline halyard-config/src/main/java/com/netflix/spinnaker/halyard/config/services/v1/BakeryService.java]: Pretty sure this line-break was accidental.

File: halyard-config/src/main/java/com/netflix/spinnaker/halyard/config/validate/v1/providers/openstack/OpenstackAccountValidator.java [hunk 16]
```diff
@@ -76,19 +112,27 @@ public void validate(ConfigProblemSetBuilder psBuilder, OpenstackAccount account
...
-          .regions(regionsList)
+          .regions(regions)
           .insecure(insecure)
+          .heatTemplateLocation(heatTemplateLocation)
+          .consulConfig(consulConfig)
           .lbaasConfig(lbaasConfig)
           .userDataFile(userDataFile)
           .build();
+      if (openstackCredentials == null) {
+        return;
```
Comment (markus-silpala-tgt) [inline halyard-config/src/main/java/com/netflix/spinnaker/halyard/config/validate/v1/providers/openstack/OpenstackAccountValidator.java]: I'm fairly certain that it's impossible for this value to be null at this point: Java constructors cannot return null, and the build() method only calls a constructor and returns its result.

Did this check come about in response to an error? Or does it simply conform to convention elsewhere within the spinnaker/halyard code base? I'm new here so this might be a silly question. @edwinavalos

File: halyard-config/src/main/java/com/netflix/spinnaker/halyard/config/validate/v1/providers/openstack/OpenstackBakeryDefaultsValidator.java [hunk 17]
```diff
@@ -0,0 +1,106 @@
...
+
+import com.netflix.spinnaker.clouddriver.google.security.GoogleNamedAccountCredentials;
+import com.netflix.spinnaker.clouddriver.openstack.security.OpenstackCredentials;
+import com.netflix.spinnaker.clouddriver.openstack.security.OpenstackNamedAccountCredentials;
+import com.netflix.spinnaker.halyard.config.model.v1.node.Validator;
+import com.netflix.spinnaker.halyard.config.model.v1.providers.openstack.OpenstackBakeryDefaults;
+import com.netflix.spinnaker.halyard.config.model.v1.providers.openstack.OpenstackBaseImage;
+import com.netflix.spinnaker.halyard.config.problem.v1.ConfigProblemSetBuilder;
+import com.netflix.spinnaker.halyard.config.validate.v1.providers.google.GoogleBaseImageValidator;
+import com.netflix.spinnaker.halyard.core.problem.v1.Problem;
```
Comment (markus-silpala-tgt) [inline halyard-config/src/main/java/com/netflix/spinnaker/halyard/config/validate/v1/providers/openstack/OpenstackBakeryDefaultsValidator.java]: Three unused imports: 
* GoogleNamedAccountCredentials
* OpenstackCredentials
* GoogleBaseImageValidator

File: halyard-deploy/src/main/java/com/netflix/spinnaker/halyard/deploy/spinnaker/v1/profile/deck/DeckProfileFactory.java [hunk 18]
```diff
@@ -124,7 +124,7 @@ protected void setProfile(Profile profile, DeploymentConfiguration deploymentCon
     if (openstackProvider.getPrimaryAccount() != null) {
       OpenstackAccount openstackAccount = (OpenstackAccount) accountService.getProviderAccount(deploymentConfiguration.getName(), "openstack", openstackProvider.getPrimaryAccount());
       //Regions in openstack are a comma separated list. Use the first as primary.
```
Comment (markus-silpala-tgt) [inline halyard-deploy/src/main/java/com/netflix/spinnaker/halyard/deploy/spinnaker/v1/profile/deck/DeckProfileFactory.java]: This comment is out of date, isn't it? Remove or update. Regions are now a list of strings from command all the way to domain objects.

**Comment by markus-silpala-tgt** [comment]
@edwinavalos I felt bad for putting in so many nit-sized remarks about unused imports, so I made a secondary PR you and use to just take all those changes en masse: https://github.com/edwinavalos/halyard/pull/6

**Comment by edwinavalos** [comment]
@lwander this one is ready for a once over again

File: halyard-cli/src/main/java/com/netflix/spinnaker/halyard/cli/command/v1/config/providers/openstack/OpenstackEditBakeryDefaultsCommand.java [hunk 19]
```diff
@@ -0,0 +1,108 @@
...
+import com.netflix.spinnaker.halyard.config.model.v1.providers.openstack.OpenstackBakeryDefaults;
+import com.netflix.spinnaker.halyard.config.model.v1.node.BakeryDefaults;
+import com.netflix.spinnaker.halyard.cli.command.v1.config.providers.bakery.AbstractEditBakeryDefaultsCommand;
+
+/**
+ * Interact with openstack provider's bakery
+ */
+@Parameters(separators = "=")
+public class OpenstackEditBakeryDefaultsCommand extends AbstractEditBakeryDefaultsCommand<OpenstackBakeryDefaults> {
+    protected String getProviderName() { return "openstack"; }
```
Comment (lwander) [inline halyard-cli/src/main/java/com/netflix/spinnaker/halyard/cli/command/v1/config/providers/openstack/OpenstackEditBakeryDefaultsCommand.java]: I'd prefer to have this on a new line to be consistent with other formatting

File: halyard-cli/src/main/java/com/netflix/spinnaker/halyard/cli/command/v1/config/providers/openstack/OpenstackEditBakeryDefaultsCommand.java [hunk 20]
```diff
@@ -0,0 +1,108 @@
...
+        required = true,
+        description = "Set the default password your images will be baked with."
+    )
+    private String password;
+
+    @Parameter(
+        names = "--insecure",
+        description = "Set the default security setting (true or false) to connect to the openstack account."
+    )
+    private boolean insecure = false;
```
Comment (lwander) [inline halyard-cli/src/main/java/com/netflix/spinnaker/halyard/cli/command/v1/config/providers/openstack/OpenstackEditBakeryDefaultsCommand.java]: This should be nullable `Boolean` to decide if it's been set or not (otherwise will always default to `false` when a user makes an unrelated edit).

File: halyard-cli/src/main/java/com/netflix/spinnaker/halyard/cli/command/v1/config/providers/openstack/OpenstackEditBaseImageCommand.java [hunk 21]
```diff
@@ -0,0 +1,72 @@
...
+        imageSettings = imageSettings != null ? imageSettings : new OpenstackBaseImage.OpenstackImageSettings();
+        baseImage.setBaseImage(imageSettings);
+        //TODO(edwinavalos) make this work for multiple virtualization settings
+        OpenstackBaseImage.OpenstackVirtualizationSettings virtualizationSettings = baseImage.getVirtualizationSettings().get(0);
+        virtualizationSettings = virtualizationSettings != null ? virtualizationSettings : new OpenstackBaseImage.OpenstackVirtualizationSettings();
+        virtualizationSettings.setSourceImageId(isSet(sourceImageId) ? sourceImageId : virtualizationSettings.getSourceImageId());
+        virtualizationSettings.setRegion(isSet(region) ? region : virtualizationSettings.getRegion());
+        virtualizationSettings.setInstanceType(isSet(instanceType) ? instanceType : virtualizationSettings.getInstanceType());
+        virtualizationSettings.setSshUserName(isSet(sshUserName) ? sshUserName : virtualizationSettings.getSshUserName());
+//        baseImage.setVirtualizationSettings(virtualizationSettings);
```
Comment (lwander) [inline halyard-cli/src/main/java/com/netflix/spinnaker/halyard/cli/command/v1/config/providers/openstack/OpenstackEditBaseImageCommand.java]: Please delete or uncomment

File: halyard-config/src/main/java/com/netflix/spinnaker/halyard/config/model/v1/providers/openstack/OpenstackProvider.java [hunk 22]
```diff
@@ -16,4 +19,19 @@ public ProviderType providerType() {
...
+
+  @Override
+  public OpenstackBakeryDefaults emptyBakeryDefaults() {
+    OpenstackBakeryDefaults result = new OpenstackBakeryDefaults();
+    // Set defaults over here
+    result.setAuthUrl(null);
+    result.setFloatingIpPool(null);
+    result.setInsecure(false);
+    result.setNetworkId(null);
+    result.setPassword(null);
```
Comment (lwander) [inline halyard-config/src/main/java/com/netflix/spinnaker/halyard/config/model/v1/providers/openstack/OpenstackProvider.java]: Setting `null` isn't needed, you can just return `new OpenstackBakeryDefaults()`

File: halyard-config/src/main/java/com/netflix/spinnaker/halyard/config/validate/v1/providers/openstack/OpenstackAccountValidator.java (hunk 16 above)
Comment (lwander) [inline halyard-config/src/main/java/com/netflix/spinnaker/halyard/config/validate/v1/providers/openstack/OpenstackAccountValidator.java]: I agree that this code path can't be hit - what's the intent here? Unless `.build()` can return null?

File: halyard-cli/src/main/java/com/netflix/spinnaker/halyard/cli/command/v1/config/providers/openstack/OpenstackCommandProperties.java (hunk 12 above)
Comment (dylanbernhardt) [inline halyard-cli/src/main/java/com/netflix/spinnaker/halyard/cli/command/v1/config/providers/openstack/OpenstackCommandProperties.java]: No, for Base Image virtualization settings it only supports one region.

**Comment by lwander** [comment]
I'm ready to merge this, any objections?

**Comment by MatthewKohner** [comment]
no objections, merge away!
